cli/deno_dir.rs

```diff
@@ -18,15 +18,62 @@ use http;
 use ring;
 use serde_json;
 use std;
+use std::collections::HashMap;
 use std::fmt::Write;
 use std::fs;
 use std::path::Path;
 use std::path::PathBuf;
 use std::result::Result;
 use std::str;
+use std::sync::Arc;
+use std::sync::Mutex;
 use url;
 use url::Url;
 
+#[derive(Clone, Default)]
+pub struct DownloadCache(
+  Arc<
+    Mutex<
+      HashMap<String, (Option<String>, Option<String>, String, Option<String>)>,
```

Comment:
This type `(Option<String>, Option<String>, String, Option<String>)` is rather obtuse. Consider defining a struct or at least an alias for it.

Reply:
Obviously... Let me tell you my idea how to solve this problem making no changes to the current test suite.

1. `fetch_remote_source_async` checks if `module_name` (which is URL 🙄 this will be addressed in #2559 ) has already been downloaded
2. if cached get result from cache and go to point 5
3. if not cached download file and follow redirects
4. cache downloaded result in `DownloadCache`
5. write down headers file for original url and redirect url
6. construct and return `ModuleMetaData` (maybe it's right place to [that TODO](https://github.com/denoland/deno/blob/master/cli/deno_dir.rs#L373-L376) and use auxiliary structure)

This approach will prevent multiple downloads of the same file on single run of deno, however it will introduce another problem - local files will be fetched from disk everytime they requested (meaning potentially during run of program you get two different versions of same file) while remote resources will be same for single run.

WDYT?

Reply:
>  local files will be fetched from disk everytime they requested

That's better than multiple downloads - sounds good. Of course this whole thing needs to get re-written, but I think your patch will help in the interim.

Do you have a test in this patch that shows it doesn't do multiple downloads anymore? I was never able to get a good test of that sort.